### PR TITLE
Issue #1036 Configure Scheduler

### PR DESCRIPTION
**Pull request by gregw**
Allows scheduler configuration of number of threads, name etc.

Signed-off-by: Greg Wilkins <gregw@webtide.com>

### CODE REVIEW

**Comment by sbordet** [comment]
@gregw no, I don't think we should budget scheduler threads. Typically there is only one.

File: jetty-util/src/main/java/org/eclipse/jetty/util/thread/ScheduledExecutorScheduler.java [hunk 1]
```diff
@@ -50,28 +55,48 @@ public ScheduledExecutorScheduler()
...
+     * @param name The name of the scheduler threads or null for automatic name
+     * @param daemon True if scheduler threads should be daemon
+     * @param classLoader The classloader to run the threads with or null to use the current thread context classloader
+     * @param threadGroup The threadgroup to use or null for no thread group
+     * @param threads The number of threads to pass to the the core {@link ScheduledThreadPoolExecutor} or -1 for a
+     * heuristic determined number of threads.
+     */
+    public ScheduledExecutorScheduler(@Name("name") String name, @Name("daemon") boolean daemon, @Name("classLoader") ClassLoader classLoader, @Name("threadGroup") ThreadGroup threadGroup, @Name("threads") int threads)
     {
         this.name = name == null ? "Scheduler-" + hashCode() : name;
```
Comment (sbordet) [inline jetty-util/src/main/java/org/eclipse/jetty/util/thread/ScheduledExecutorScheduler.java]: Now that the name can be specified in XML, when using an empty tag the empty string will be passed in, so I think this should be improved using `StringUtil.isBlank()`.

File: jetty-util/src/main/java/org/eclipse/jetty/util/thread/ScheduledExecutorScheduler.java [hunk 2]
```diff
@@ -50,28 +55,48 @@ public ScheduledExecutorScheduler()
...
+        this.classloader = classLoader == null ? Thread.currentThread().getContextClassLoader() : classLoader;
         this.threadGroup = threadGroup;
+        this.threads = threads;
     }
 
     @Override
     protected void doStart() throws Exception
     {
-        scheduler = new ScheduledThreadPoolExecutor(1, r ->
+        int size = threads > 0 ? threads : Math.max(1, ProcessorUtils.availableProcessors() / 4);
```
Comment (sbordet) [inline jetty-util/src/main/java/org/eclipse/jetty/util/thread/ScheduledExecutorScheduler.java]: I don't see any reason to have more than 1 thread by default, like it was previously. in large machines we are going to possibly waste tens of threads.

**Comment by rrorden** [comment]
The changes merged for this issue have uncovered what I believe is a bug in the code for the XmlConfiguration$JettyXmlConfiguration.construct(Class<?> klass, Object[] arguments, Map<String, Object> namedArgMap) method. 

I ran into this while testing jetty 9.4.26 on the Perc Virtual Machine (PVM), a real-time commercial JVM that has been available since 1996 and which I help maintain. Perc is an alternative JVM like IBM J9 and others. We have customers using it with Jetty, Tomcat and other frameworks.

The modifications for issue #1036 and pull request #4086  include adding <Arg> definitions to jetty.xml for constructing a ScheduledExecutorScheduler and adding parameter annotations to that constructor in order to match the <Arg> specified in jetty.xml:
https://github.com/eclipse/jetty.project/commit/92e4d73dcb69266984f5a566fe8885cf3ab8f733

It also adds a new three-argument constructor to ScheduledExecutorScheduler:
public ScheduledExecutorScheduler(String name, boolean daemon, ClassLoader classLoader)

The problem is that XmlConfiguration$JettyXmlConfiguration.construct() calls Class.getConstructors() on ScheduledExecutorScheduler and then iterates through them until it finds one with the same number of parameters as the length of the passed in arguments array (3 in this case). Then it calls Constructor.getParameterAnnotations() in preparation to create a "swizzled" array of arguments that match up to the namedArgMap that was passed into the method. Unfortunately, the Java spec for Class.getConstructors() does not guarantee the order of the returned constructors. OpenJDK returns the newly parameterized (String, boolean, int) constructor first, but Perc happens to return the new (String, boolean, ClassLoader) constructor first and Constructor.getParameterAnnotations() returns an empty array. This causes construct() to debug-log "Invoking constructor, no parameter annotations" and invoke the wrong constructor. 

If the intent is to be able to call a constructor with the xml-specified parameter types using annotations, then I believe the construct() method needs to keep searching for a constructor that a) has parameter annotations and b) those annotations match the namedArgMap elements. It certainly is wrong to select the constructor based solely on the number of parameters it takes.

**Comment by sbordet** [comment]
@rrorden thanks for reporting this. Can you please open a new issue with your comment above?

**Comment by rrorden** [comment]
> @rrorden thanks for reporting this. Can you please open a new issue with your comment above?

@gregw opened #4550 to address this issue. I was able to download and build a snapshot with his changes he made and now it works with the Perc JVM. Amazing response time! Thanks!

**Pull request merged**